core/src/main/java/io/grpc/Metadata.java

```diff
@@ -397,18 +414,60 @@ private String toStringInternal() {
     }
 
     private final String name;
-    private final byte[] asciiName;
+
+    private final String normalName;
+    private final byte[] normalNameBytes;
+
+    private static BitSet generateValidTChars() {
+      BitSet valid  = new BitSet(0x7f);
+      valid.set('-');
+      valid.set('_');
+      for (char c = '0'; c <= '9'; c++) {
+        valid.set(c);
+      }
+      // Only validates after normalization, so we exclude uppercase.
+      for (char c = 'a'; c <= 'z'; c++) {
+        valid.set(c);
+      }
+      return valid;
+    }
+
+    private static String validateName(String n) {
+      checkNotNull(n);
+      checkArgument(n.length() != 0, "token must have at least 1 tchar");
+      for (int i = 0; i < n.length(); i++) {
+        char tChar = n.charAt(i);
+        // TODO(notcarl): remove this hack once pseudo headers are properly handled
+        if (tChar == ':' && i == 0) {
+          continue;
+        }
+
+        checkArgument(VALID_T_CHARS.get(tChar), "Invalid character '%s' in key name", tChar);
+      }
+      return n;
+    }
 
     private Key(String name) {
-      this.name = Preconditions.checkNotNull(name, "name").toLowerCase(Locale.ROOT).intern();
-      this.asciiName = this.name.getBytes(US_ASCII);
+      this.name = checkNotNull(name);
+      this.normalName = validateName(this.name.toLowerCase(Locale.ROOT));
+      this.normalNameBytes = this.normalName.getBytes(US_ASCII);
     }
 
+    /**
+     * @return The original name used to create this key.
+     */
     public String name() {
       return name;
     }
 
     /**
+     * @return The normalized name for this key.
+     */
+    public String normalName() {
```

Comment:
Instead of creating this, we could just not change the API and have `name()` be normalized. I can't think of legitimate code that should call `name()` instead of `normalName()`, except maybe logging. Because of that, if we really _need_ to have the original name we add `originalName()` or some such.